consensus/core/src/linearizer.rs

```diff
@@ -124,38 +147,75 @@ impl Linearizer {
         let mut to_commit = Vec::new();
         assert!(committed.insert(leader_block_ref));
 
-        while let Some(x) = buffer.pop() {
-            to_commit.push(x.clone());
-
-            let ancestors: Vec<VerifiedBlock> = dag_state
-                .get_blocks(
-                    &x.ancestors()
-                        .iter()
-                        .copied()
-                        .filter(|ancestor| {
-                            // We skip the block if we already committed it or we reached a
-                            // round that we already committed.
-                            // TODO: for Fast Path we need to ammend the recursion rule here and allow us to commit blocks all the way up to the `gc_round`.
-                            // Some additional work will be needed to make sure that we keep the uncommitted blocks up to the `gc_round` across commits.
-                            !committed.contains(ancestor)
-                                && last_committed_rounds[ancestor.author] < ancestor.round
-                        })
-                        .filter(|ancestor| {
-                            // Keep the block if GC is not enabled or it is enabled and the block is above the gc_round. We do this
-                            // to stop the recursion early and avoid going to deep when it's unnecessary.
-                            !gc_enabled || ancestor.round > gc_round
-                        })
-                        .collect::<Vec<_>>(),
-                )
-                .into_iter()
-                .map(|ancestor_opt| {
-                    ancestor_opt.expect("We should have all uncommitted blocks in dag state.")
-                })
-                .collect();
-
-            for ancestor in ancestors {
-                buffer.push(ancestor.clone());
-                assert!(committed.insert(ancestor.reference()));
+        // The new logic will perform the recursion without stopping at the highest round round that has been committed per authority. Instead it will
+        // continue until gc_round is reached.
+        if context
+            .protocol_config
+            .consensus_linearizer_collect_subdag_v2()
+        {
```

Comment:
Decided to split completely the new/previous logic to make things safer. Will clean up after deploy.

Reply:
Not a big deal as you will clean it up, but from a diff perspective changing the two spots alone that use the new behavior does seem safer/easier to review  and you can keep the comments.

Reply:
Yeah I thought that it would be cleaner to review the logic as whole separately before/after rather than having to navigate through various points that enable different code paths. But if you believe it's tougher to review then I am happy to reconsider on this.